llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp

```diff
@@ -5869,6 +5869,12 @@ bool SelectionDAG::isKnownNeverNaN(SDValue Op, bool SNaN,
                            ? APInt::getAllOnes(VT.getVectorNumElements())
                            : APInt(1, 1);
 
+  // If all users of this operand is annotated with nnan, we can assume
+  // this operand is not NaN, since nnan also affects inputs.
+  if (llvm::all_of(Op->users(),
+                   [](const SDNode *N) { return N->getFlags().hasNoNaNs(); }))
+    return true;
```

Comment:
We should get fast-math flags from somewhere.
```
def : GCNPat<
    (fcanonicalize (vt is_canonicalized:$src)),
    (COPY vt:$src)
  >;
```
For this pattern, when `fcanonicalize` is called with `nnan` , `isKnownNeverNaN` should return true directly for this case, rather than check all users.

Reply:
It would break the test `test73_nnan` in `llvm/test/CodeGen/AMDGPU/combine_andor_with_cmps.ll`. But drop it for now is fine.

Reply:
that seems like it should be fixed by a more specific use context combine